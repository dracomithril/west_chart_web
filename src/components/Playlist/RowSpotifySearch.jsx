/**
 * Created by XKTR67 on 5/11/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { MenuItem } from '@material-ui/core/Menu';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { faCaretDown, faCaretUp, faSearch, faSync, faTimes } from '@fortawesome/fontawesome-free-solid';
import './playlist.css';
import TrackPreview from './TrackPreview';

class RowSpotifySearch extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showList: false,
    };
  }

  render() {
    const {
      id,
      artist,
      title,
      full_title,
      selected: selected_track,
      items = [],
      onSwap,
      onUpdateClick,
      onSearchClick,
      onClearClick,
    } = this.props;
    const { showList } = this.state;
    const tracks_list = items.map(track => (
      <MenuItem
        key={track.id}
        id={`mi_select_track_${id}`}
        onClick={() => {
          onUpdateClick({ id, value: track, field: 'selected' });
          this.setState({ showList: false });
        }}
        style={{ backgroundColor: 'snow', height: 'unset' }}
      >
        <TrackPreview
          artists={track.artists}
          external_urls={track.external_urls}
          preview_url={track.preview_url}
          trackName={track.name}
          noLink
        />
      </MenuItem>
    ));

    const condition = selected_track && selected_track.preview_url !== undefined;
    const haveIssue = !artist || !title;

    const selectTrackClass = selected_track ? 'row-spotify-search__root--good' : 'row-spotify-search__root--error';
    const trackSearchClass = haveIssue
      ? 'row-spotify-search__track-search--error'
      : 'row-spotify-search__track-search--good';
    return (
      <div className={selectTrackClass}>
        <span>{full_title || 'No Title'}</span>
        <div>
          <div>
            <div className={trackSearchClass}>
              <label htmlFor={`${id}_artist`}>
                artist:
                <input
                  type="text"
                  id={`${id}_artist`}
                  value={artist || ''}
                  placeholder="artist name"
                  onChange={({ target }) => {
                    onUpdateClick({ id, value: target.value, field: 'artist' });
                  }}
                />
              </label>
              <label htmlFor={`${id}_title`}>
                title:
                <input
                  type="text"
                  id={`${id}_title`}
                  value={title || ''}
                  placeholder="song title"
                  onChange={({ target }) => onUpdateClick && onUpdateClick({ id, value: target.value, field: 'title' })}
                />
              </label>
              <div className="row-spotify-search__button-group">
                <button type="button" onClick={() => onSwap && onSwap(id)} title="swap artist with title">
                  <FontAwesomeIcon icon={faSync} />
                </button>
                <button
                  type="button"
                  id={`button-${id}`}
                  onClick={() => onSearchClick && onSearchClick()}
                  title="search for tracks"
                >
                  <FontAwesomeIcon icon={faSearch} />
                </button>
                <button type="button" title="clear selected" onClick={() => onClearClick && onClearClick({ id })}>
                  <FontAwesomeIcon icon={faTimes} />
                </button>
                {tracks_list.length !== 0 && (
                  <button
                    type="button"
                    onClick={() => {
                      this.setState({ showList: !showList });
                    }}
                    title="show/hide found tracks"
                  >
                    {tracks_list.length !== 0 && <FontAwesomeIcon icon={showList ? faCaretUp : faCaretDown} />}
                  </button>
                )}
              </div>
            </div>
            {showList && <ol>{tracks_list}</ol>}
          </div>
          {condition && (
            <div className="row-spotify-search__track_view">
              <TrackPreview
                artists={selected_track.artists}
                external_urls={selected_track.external_urls}
                preview_url={selected_track.preview_url}
                trackName={selected_track.name}
              />
            </div>
          )}
        </div>
      </div>
    );
  }
}

RowSpotifySearch.contextTypes = {
  store: PropTypes.shape,
};

const trackProps = PropTypes.shape({
  id: PropTypes.string,
  artist: PropTypes.string,
  preview_url: PropTypes.string,
  external_urls: PropTypes.object,
  name: PropTypes.string,
});

RowSpotifySearch.propTypes = {
  onSwap: PropTypes.func,
  id: PropTypes.string,
  items: PropTypes.arrayOf(trackProps),
  artist: PropTypes.string,
  title: PropTypes.string,
  full_title: PropTypes.string,
  selected: trackProps,
  onUpdateClick: PropTypes.func,
  onSearchClick: PropTypes.func,
  onClearClick: PropTypes.func,
};
RowSpotifySearch.defaultProps = {
  items: [],
};
export default RowSpotifySearch;
