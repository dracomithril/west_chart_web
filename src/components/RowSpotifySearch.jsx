/**
 * Created by XKTR67 on 5/11/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonGroup, DropdownButton, MenuItem } from 'react-bootstrap';
import { searchForMusic } from '../utils/spotify_utils';
import './components.css';

import action_types from './../reducers/action_types';

const TrackPreview = props => {
  const { track, nolink } = props;
  const artists = track.artists.map(elem => elem.name).join(' & ');
  const audio =
    track.preview_url !== null ? (
      <audio controls>
        <source src={track.preview_url} type="audio/mpeg" />
      </audio>
    ) : (
      <span style={{ color: 'red' }}>
        No preview
        {!nolink && (
          <a href={(track.external_urls || {}).spotify} target="_newtab">
            {' go to '}
            <i className="fab fa-spotify" aria-hidden="true" />
          </a>
        )}
      </span>
    );
  return (
    <div className="track_view">
      <strong>{track.name}</strong>
      <br />
      {audio}
      <div>
        <span>by:</span>
        <span className="artist_name">{artists}</span>
      </div>
    </div>
  );
};
TrackPreview.propTypes = {
  track: PropTypes.object,
  nolink: PropTypes.bool,
};

const RowSpotifySearch = ({ search_elem }, { store }) => {
  const create_menuItems = track => {
    const selectTrack = () => {
      console.info('track was selected %j', track);
      store.dispatch({
        type: action_types.UPDATE_SINGLE_SEARCH,
        field: 'selected',
        value: track,
        id: search_elem.search_id,
      });
    };
    return (
      <MenuItem key={track.id} id={`mi_select_track_${search_elem.id}`} onClick={selectTrack}>
        <TrackPreview track={track} nolink />
      </MenuItem>
    );
  };
  const items = (search_elem.items || []).map(create_menuItems);

  const update_artist = ({ target }) => {
    store.dispatch({
      type: action_types.UPDATE_SINGLE_SEARCH,
      field: 'artist',
      value: target.value,
      id: target.id,
    });
  };
  const update_title = ({ target }) => {
    store.dispatch({
      type: action_types.UPDATE_SINGLE_SEARCH,
      field: 'title',
      value: target.value,
      id: target.id,
    });
  };
  const selected_track = search_elem.selected;
  const condition = selected_track !== undefined && selected_track.preview_url !== undefined;
  const haveIssue = !search_elem.artist || !search_elem.title;
  return (
    <tr style={condition ? undefined : { background: 'red' }}>
      <td>
        <div className={`track_view ${haveIssue ? 'error_bg' : ''}`}>
          <span>{search_elem.full_title}</span>
          <br />
          <input
            type="text"
            id={search_elem.search_id}
            value={search_elem.artist || ''}
            onChange={update_artist}
          />
          <button
            onClick={() => {
              store.dispatch({
                type: action_types.SWAP_FIELDS,
                id: search_elem.search_id,
              });
            }}
          >
            <i className="fas fa-sync" aria-hidden="true" />
          </button>
          <input
            type="text"
            id={search_elem.search_id}
            value={search_elem.title || ''}
            onChange={update_title}
          />
        </div>
      </td>
      <td>
        <div>
          <ButtonGroup vertical>
            <Button
              id={`button-${search_elem.id}`}
              onClick={() => {
                searchForMusic(search_elem, store).then(res =>
                  store.dispatch({
                    type: action_types.UPDATE_SINGLE_SEARCH,
                    field: 'items',
                    value: res.value,
                    id: res.id,
                  }),
                );
              }}
              bsStyle="info"
            >
              search
            </Button>
            <DropdownButton
              disabled={items.length === 0}
              title="select"
              key={search_elem.search_id}
              id={`dropdown-basic-${search_elem.search_id}`}
              bsStyle={items.length === 0 ? 'warning' : 'success'}
            >
              {items}
            </DropdownButton>
          </ButtonGroup>
        </div>
      </td>
      <td>{condition && <TrackPreview track={selected_track} />}</td>
    </tr>
  );
};
RowSpotifySearch.contextTypes = {
  store: PropTypes.object,
};
RowSpotifySearch.propTypes = {
  search_elem: PropTypes.object,
};
export default RowSpotifySearch;
