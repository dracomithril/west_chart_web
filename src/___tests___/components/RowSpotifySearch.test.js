/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import RowSpotifySearch from './../../components/RowSpotifySearch';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../data/initial_state.json');

const mockStore = configureMockStore([]);
describe('<RowSpotifySearch/>', () => {
  it('renders without crashing', () => {
    const store = mockStore(initial_state);
    let newVar = { id: 0, items: [], selected: [], search_id: 0 };
    const wrapper = shallow(
      <RowSpotifySearch search_elem={newVar}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
