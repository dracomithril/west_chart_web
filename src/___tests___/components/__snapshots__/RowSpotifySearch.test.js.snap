// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RowSpotifySearch/> renders without crashing 1`] = `
<tr
  style={
    Object {
      "background": "red",
    }
  }
>
  <td>
    <div
      className="track_view error_bg"
    >
      <span />
      <br />
      <input
        id={0}
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-refresh"
        />
      </button>
      <input
        id={0}
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </td>
  <td>
    <div>
      <ButtonGroup
        block={false}
        bsClass="btn-group"
        justified={false}
        vertical={true}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="info"
          disabled={false}
          id="button-0"
          onClick={[Function]}
        >
          search
        </Button>
        <DropdownButton
          bsStyle="warning"
          disabled={true}
          id="dropdown-basic-0"
          key="0"
          title="select"
        />
      </ButtonGroup>
    </div>
  </td>
  <td />
</tr>
`;
